src/corehost/cli/fxr/command_line.cpp

```diff
@@ -0,0 +1,337 @@
+        /*out*/ int *new_argoff,
+        /*out*/ pal::string_t &app_candidate,
+        /*out*/ opt_map_t &opts)
+    {
+        std::vector<known_options> known_opts = get_known_opts(exec_mode, mode);
+
+        // Parse the known arguments if any.
+        int num_parsed = 0;
+        if (!parse_known_args(argc - argoff, &argv[argoff], known_opts, &opts, &num_parsed))
+        {
+            trace::error(_X("Failed to parse supported options or their values:"));
+            for (const auto& opt : known_opts)
+            {
+                const host_option &arg = get_host_option(opt);
+                trace::error(_X("  %-37s  %s"), (arg.option + _X(" ") + arg.argument).c_str(), arg.description.c_str());
+            }
+            return StatusCode::InvalidArgFailure;
+        }
+
+        app_candidate = host_info.app_path;
+        *new_argoff = argoff + num_parsed;
+        bool doesAppExist = false;
+        if (mode == host_mode_t::apphost)
+        {
+            doesAppExist = pal::realpath(&app_candidate);
+        }
+        else
+        {
+            trace::verbose(_X("Using the provided arguments to determine the application to execute."));
+            if (*new_argoff >= argc)
+            {
+                command_line::print_muxer_usage(!is_sdk_dir_present(host_info.dotnet_root));
+                return StatusCode::InvalidArgFailure;
+            }
+
+            app_candidate = argv[*new_argoff];
+
+            bool is_app_managed = ends_with(app_candidate, _X(".dll"), false) || ends_with(app_candidate, _X(".exe"), false);
+            if (!is_app_managed)
+            {
+                trace::verbose(_X("Application '%s' is not a managed executable."), app_candidate.c_str());
+                if (!exec_mode)
+                {
+                    // Route to CLI.
+                    return StatusCode::AppArgNotRunnable;
+                }
+            }
+
+            doesAppExist = pal::realpath(&app_candidate);
+            if (!doesAppExist)
+            {
+                trace::verbose(_X("Application '%s' does not exist."), app_candidate.c_str());
+                if (!exec_mode)
+                {
+                    // Route to CLI.
+                    return StatusCode::AppArgNotRunnable;
+                }
+            }
+
+            if (!is_app_managed && doesAppExist)
+            {
+                assert(exec_mode == true);
+                trace::error(_X("dotnet exec needs a managed .dll or .exe extension. The application specified was '%s'"), app_candidate.c_str());
+                return StatusCode::InvalidArgFailure;
+            }
+        }
+
+        // App is managed executable.
+        if (!doesAppExist)
+        {
+            trace::error(_X("The application to execute does not exist: '%s'"), app_candidate.c_str());
+            return StatusCode::InvalidArgFailure;
+        }
+
+        return StatusCode::Success;
+    }
+}
+
+pal::string_t command_line::get_last_known_arg(
```

Comment:
Nit: I find this name confusing. Maybe something like `get_know_option_value`. or just `get_option_value` (since it takes `known_options` enum so that would be redundant).